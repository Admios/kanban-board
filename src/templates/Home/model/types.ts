export interface Todo {
  id: string;
  stateId: string;
  text: string;
  position: number;
  owner: string;
}

export interface State {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
