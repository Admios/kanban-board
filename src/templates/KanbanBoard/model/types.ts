export interface Task {
  id: string;
  stateId: string;
  text: string;
  position: number;
}

export interface State {
  id: string;
  boardId: string;
  name: string;
  position: number;
  color: string;
}

export interface Board {
  id: string;
  name: string;
  owner: string;
}
