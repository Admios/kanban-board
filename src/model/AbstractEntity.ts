/**
 * Entities represent how the data is stored in the database.
 */
export interface AbstractEntity extends Record<string, any> {}
