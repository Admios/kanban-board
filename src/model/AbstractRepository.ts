export interface AbstractRepository<T> {
  findById(id: string): Promise<T>;
  list(): Promise<T[]>;
  create(entity: T): Promise<T>;
  update(entity: T): Promise<T>;
  delete(entity: T): Promise<T>;
}
