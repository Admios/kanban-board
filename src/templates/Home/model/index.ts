import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable, produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";

export type { State, Task } from "./types";

type HomeState = Immutable<{
  tasks: Record<string, Task[]>;
  states: Record<string, State>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTasks: TaskDTO[],
    initialStates: StateDTO[],
    initialUser?: UserDTO,
  ): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates, initialUser) {
    const statesMap: Record<string, State> = {};
    const tasksMap: Record<string, Task[]> = {};

    initialStates.forEach((backendState) => {
      const result: State = {
        ...backendState,
        id: backendState.id,
        position: backendState.position,
      };
      statesMap[result.id] = result;
    });

    initialTasks.forEach((backendTask) => {
      const { stateId } = backendTask;
      if (!tasksMap[stateId]) {
        tasksMap[stateId] = [];
      }
      tasksMap[stateId][backendTask.position] = {
        ...backendTask,
        stateId,
        id: backendTask.id,
        position: backendTask.position,
      };
    });

    set({ tasks: tasksMap, states: statesMap, user: initialUser });
  },

  addTask: (newTask) => {
    const result = { ...newTask };
    const tasks = produce(get().tasks, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }

      draft[newTask.stateId].push(result);
    });

    set({ tasks });
    return result;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const tasks = produce(get().tasks, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task.id !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask);
      } else {
        destinationState.push(newTask);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasks });
  },

  addState: (newState) => {
    const currentModel = get();
    const tasks = produce(currentModel.tasks, (draft) => {
      draft[newState.id] = [];
    });
    const states = produce(currentModel.states, (draft) => {
      const position = Object.keys(currentModel.states).length;
      draft[newState.id] = { ...newState, position };
    });

    set({ states, tasks });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task.id === taskId);
        if (taskIndex > -1) {
          stateTasks[taskIndex] = {
            ...stateTasks[taskIndex],
            ...updatedValues,
          };
          break;
        }
      }
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const currentState = get();
    const tasks = produce(currentState.tasks, (draft) => {
      for (const stateId in draft) {
        const stateTasks = draft[stateId];
        const taskIndex = stateTasks.findIndex((task) => task?.id === taskId);
        if (taskIndex > -1) {
          stateTasks.splice(taskIndex, 1);
          break;
        }
      }
    });

    set({ tasks });
  },
});

export const useZustand = create(stateCreator);
