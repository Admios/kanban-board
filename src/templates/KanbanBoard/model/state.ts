import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
