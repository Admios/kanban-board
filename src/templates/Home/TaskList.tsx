import { Box, Flex, Heading, useDisclosure } from "@chakra-ui/react";
import { useMemo, useState } from "react";
import { AddColumnModal } from "./AddColumnModal";
import { Column } from "./Column";
import { Header } from "./Header";
import { Column as ColumnType, useZustand } from "./state";
import { AddTodoModal } from "./AddTodoModal";

export function TaskList() {
  const addColumn = useZustand((store) => store.addColumn);
  const addTodo = useZustand((store) => store.addTodo);
  const columns = useZustand((store) => store.columns);
  const [addTodoModalColId, setTodoModalColId] = useState<string | undefined>();
  const {
    isOpen: isColumnDialogOpen,
    onOpen: onOpenColumnDialog,
    onClose: onCloseColumnDialog,
  } = useDisclosure();

  const sortedColumns = useMemo(() => {
    return Object.values(columns).sort(
      (valueA, valueB) => valueA.position - valueB.position,
    );
  }, [columns]);

  const handleCreateRandomTasks = () => {
    const randomTasks = new Set<string>();
    let firstColumn: ColumnType;

    if (!sortedColumns.length) {
      firstColumn = addColumn({
        name: "Random Column",
        backendId: null,
        color: "black",
      });
    } else {
      firstColumn = sortedColumns[0];
    }

    while (randomTasks.size < 10) {
      randomTasks.add(`Random Task ${Math.floor(Math.random() * 100)}`);
    }

    randomTasks.forEach((task) =>
      addTodo({ text: task, columnId: firstColumn.id, backendId: null }),
    );
  };

  return (
    <Box margin="4">
      <Header
        handleCreateRandomTasks={handleCreateRandomTasks}
        onOpenColumnDialog={onOpenColumnDialog}
      />
      <Heading mx="auto" paddingBottom="2">
        Board
      </Heading>

      <Flex direction={"row"} gap="2">
        {sortedColumns.map((value) => (
          <Column
            key={value.name}
            colId={value.id}
            colTitle={value.name}
            color={value.color}
            onOpenCreateTodoModal={() => setTodoModalColId(value.id)}
          />
        ))}
      </Flex>

      <AddColumnModal
        isOpen={isColumnDialogOpen}
        onClose={onCloseColumnDialog}
      />
      <AddTodoModal
        isOpen={!!addTodoModalColId}
        onClose={() => setTodoModalColId(undefined)}
        columnId={addTodoModalColId}
      />
    </Box>
  );
}
