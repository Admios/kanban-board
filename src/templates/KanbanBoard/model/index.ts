import { produce } from "immer";
import { StateCreator, create } from "zustand";
import { State, Task } from "./types";
import { KanbanState, KanbanActions } from "./state";

export type { State, Task } from "./types";

const stateCreator: StateCreator<KanbanState & KanbanActions> = (set, get) => ({
  statesOrder: [],
  tasksOrder: {},
  tasks: {},
  states: {},

  initialize(initialTasks, initialStates) {
    const statesOrderSparseArray: string[] = [];
    const tasksOrder: Record<string, string[]> = {};
    const states: Record<string, State> = {};
    const tasks: Record<string, Task> = {};

    initialTasks.forEach((backendTask) => {
      tasks[backendTask.id] = backendTask;
      if (!tasksOrder[backendTask.stateId]) {
        tasksOrder[backendTask.stateId] = [];
      }
      // Warning: saved as sparse array! Compression is done later
      tasksOrder[backendTask.stateId][backendTask.position] = backendTask.id;
    });

    initialStates.forEach((backendState) => {
      states[backendState.id] = backendState;
      // Warning: saved as sparse array! compressed later
      statesOrderSparseArray[backendState.position] = backendState.id;
      // Compress sparse array
      tasksOrder[backendState.id] = (tasksOrder[backendState.id] ?? []).filter(
        (id) => !!id,
      );
    });

    // Compress sparse array
    const statesOrder = statesOrderSparseArray.filter((id) => !!id);

    set({
      statesOrder,
      tasksOrder,
      tasks,
      states,
    });
  },

  setUser(user) {
    set({ user });
  },

  addTask: (newTask) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[newTask.stateId] = newTask;
    });

    const tasksOrder = produce(get().tasksOrder, (draft) => {
      if (!draft[newTask.stateId]) {
        draft[newTask.stateId] = [];
      }
      draft[newTask.stateId].push(newTask.id);
    });

    set({ tasks, tasksOrder });
    return newTask;
  },

  moveTask: (newTask, fromStateId, toStateId, position) => {
    const newTasksOrder = produce(get().tasksOrder, (draftState) => {
      const sourceState = (draftState[fromStateId] ?? []).filter(
        (task) => task !== newTask.id,
      );

      let destinationState = draftState[toStateId] ?? [];
      if (fromStateId === toStateId) {
        destinationState = sourceState;
      }

      if (position <= destinationState.length) {
        destinationState.splice(position, 0, newTask.id);
      } else {
        destinationState.push(newTask.id);
      }

      draftState[fromStateId] = sourceState;
      draftState[toStateId] = destinationState;
    });

    set({ tasksOrder: newTasksOrder });
  },

  addState: (newState) => {
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[newState.id] = [];
    });

    const statesOrder = produce(get().statesOrder, (draft) => {
      draft.push(newState.id);
    });

    const states = produce(get().states, (draft) => {
      draft[newState.id] = newState;
    });

    set({ states, tasksOrder, statesOrder });
    return states[newState.id];
  },

  editTask: (taskId, updatedValues) => {
    const tasks = produce(get().tasks, (draft) => {
      draft[taskId] = { ...draft[taskId], ...updatedValues };
    });

    set({ tasks });
  },

  deleteTask: (taskId) => {
    const { stateId } = get().tasks[taskId];
    const tasksOrder = produce(get().tasksOrder, (draft) => {
      draft[stateId] = draft[stateId].filter((id) => id !== taskId);
    });

    const tasks = produce(get().tasks, (draft) => {
      delete draft[taskId];
    });

    set({ tasks, tasksOrder });
  },
});

export const useZustand = create(stateCreator);
