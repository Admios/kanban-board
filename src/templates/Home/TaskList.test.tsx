import { act, render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TaskList } from "./TaskList";
import { useZustand } from "./model";

jest.mock("./clearCookies.ts");
jest.mock("./homeServerActions.ts");

afterEach(() => {
  useZustand.setState({
    states: {},
    tasks: {},
    user: {
      username: "test",
    },
  });
});

function setInitialState() {
  // NOTE: The state's positions are not in order!
  // They should get sorted by the component itself
  act(() => {
    useZustand.setState({
      states: {
        "state-1": {
          id: "state-1",
          name: "To do",
          position: 100,
          color: "red",
          owner: "test",
        },
        "state-2": {
          id: "state-2",
          name: "In progress",
          position: 0,
          color: "blue",
          owner: "test",
        },
        "state-3": {
          id: "state-3",
          name: "Done",
          position: 3,
          color: "green",
          owner: "test",
        },
      },
      tasks: {
        "state-1": [],
        "state-2": [],
        "state-3": [],
      },
    });
  });
}

it("should render the sorted states", () => {
  setInitialState();

  const { container } = render(<TaskList />);
  expect(container).toMatchSnapshot("Default Home Page");

  ["In progress", "To do", "Done"].forEach((stateTitle) => {
    const element = screen.queryByText(stateTitle);
    expect(element).toBeInTheDocument();
  });
});

// it should render when empty
it("should render when empty", () => {
  const { container } = render(<TaskList />);
  expect(container).toMatchSnapshot("Empty Home Page");

  ["In progress", "To do", "Done"].forEach((stateTitle) => {
    const element = screen.queryByText(stateTitle);
    expect(element).not.toBeInTheDocument();
  });
});

it("should open the AddStateModal when button is pressed", async () => {
  render(<TaskList />);

  // Click the button
  const button = screen.getByRole("button", { name: "Add State" });
  expect(button).toBeInTheDocument();

  await userEvent.click(button);

  const modal = screen.getByRole("dialog");
  modal.setAttribute("style", ""); // This value has animation! We don't want that in our snapshot
  expect(modal).toMatchSnapshot("AddStateModal");
});

describe("sould create 10 random tasks when button is pressed", () => {
  it("should add new tasks on the first state", async () => {
    setInitialState();
    render(<TaskList />);

    const createRandomTasksButton = screen.getByRole("button", {
      name: "Create 10 random tasks",
    });
    expect(createRandomTasksButton).toBeInTheDocument();

    const state = screen.getByTitle("In progress");
    expect(state).toBeInTheDocument();

    expect(within(state).queryAllByTitle("task")).toHaveLength(0);

    await userEvent.click(createRandomTasksButton);

    await waitFor(() =>
      expect(within(state).getAllByTitle("task")).toHaveLength(10),
    );
  });

  it("should add new tasks on a new state", async () => {
    render(<TaskList />);

    const createRandomTasksButton = screen.getByRole("button", {
      name: "Create 10 random tasks",
    });
    expect(createRandomTasksButton).toBeInTheDocument();

    let randomState: HTMLElement | null;

    randomState = screen.queryByTitle("Random State");
    expect(randomState).not.toBeInTheDocument();

    await userEvent.click(createRandomTasksButton);

    randomState = screen.getByTitle("Random State");
    await waitFor(() => expect(randomState).toBeInTheDocument());
    expect(within(randomState).getAllByTitle("task")).toHaveLength(10);
  });
});
