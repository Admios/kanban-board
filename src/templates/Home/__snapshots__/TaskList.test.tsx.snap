// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should open the AddColumnModal when button is pressed: AddColumnModal 1`] = `
<section
  aria-describedby="chakra-modal--body-:r3:"
  aria-labelledby="chakra-modal--header-:r3:"
  aria-modal="true"
  class="chakra-modal__content css-6yiygt"
  id="chakra-modal-:r3:"
  role="dialog"
  style=""
  tabindex="-1"
>
  <header
    class="chakra-modal__header css-72fd9l"
    id="chakra-modal--header-:r3:"
  >
    Add Column
  </header>
  <button
    aria-label="Close"
    class="chakra-modal__close-btn css-6sdaxy"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="chakra-icon css-onkibi"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
        fill="currentColor"
      />
    </svg>
  </button>
  <div
    class="chakra-modal__body css-0"
    id="chakra-modal--body-:r3:"
  >
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        data-invalid=""
        for="field-:r6:"
        id="field-:r6:-label"
      >
        Name:
      </label>
      <input
        aria-describedby="field-:r6:-feedback"
        aria-invalid="true"
        class="chakra-input css-0"
        id="field-:r6:"
        type="text"
        value=""
      />
      <div
        aria-live="polite"
        class="chakra-form__error-message css-70qvj9"
        id="field-:r6:-feedback"
      >
        Name is required.
      </div>
    </div>
  </div>
  <footer
    class="chakra-modal__footer css-192qrng"
  >
    <button
      class="chakra-button css-eoelga"
      type="button"
    >
      Close
    </button>
    <button
      class="chakra-button css-eoelga"
      disabled=""
      type="button"
    >
      Add Column
    </button>
  </footer>
</section>
`;

exports[`should render: Default Home Page 1`] = `
<div>
  <div
    class="css-0"
  >
    <header
      class="css-0"
    >
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Logout
      </button>
      <button
        class="chakra-button css-18zcl9e"
        type="button"
      >
        Add Column
      </button>
      <h2
        class="chakra-heading css-ldihrc"
      >
        Board
      </h2>
    </header>
    <div
      class="css-1xhj18k"
    />
  </div>
</div>
`;
